File: src/lib/identity.ts
import type { Player } from "../../shared/types.ts";

const KEY = "nflpool.players.v1";
const LEGACY_KEY = "nflpool.player.v1";

/** A name this device can pick as, and the token that proves it. */
export interface Identity extends Player {
  /** Absent on devices that signed in before tokens existed, or when only the cookie holds it. */
  token?: string;
  /** True when the commissioner put this person on this device (a parent picking for the family). */
  managed?: boolean;
}

interface Store {
  activeId: string | null;
  people: Identity[];
}

const EMPTY: Store = { activeId: null, people: [] };

function read(): Store {
  try {
    const raw = localStorage.getItem(KEY);
    if (raw) {
      const s = JSON.parse(raw) as Partial<Store>;
      const people = Array.isArray(s.people) ? s.people.filter(isIdentity) : [];
      return { activeId: typeof s.activeId === "string" ? s.activeId : (people[0]?.id ?? null), people };
    }
    // One identity per device was the old shape; carry it over.
    const legacy = localStorage.getItem(LEGACY_KEY);
    if (legacy) {
      const p = JSON.parse(legacy) as Partial<Identity>;
      if (isIdentity(p)) return { activeId: p.id, people: [p] };
    }
  } catch {
    /* private mode, corrupt JSON */
  }
  return EMPTY;
}

function isIdentity(p: unknown): p is Identity {
  const v = p as Partial<Identity> | null;
  return !!v && typeof v.id === "string" && typeof v.name === "string";
}

function write(store: Store): void {
  try {
    localStorage.setItem(KEY, JSON.stringify(store));
    localStorage.removeItem(LEGACY_KEY);
  } catch {
    /* private mode etc. */
  }
}

export const loadStore = read;

/** Who this device is picking as right now. The API client reads this on every request. */
export function loadPlayer(): Identity | null {
  const { activeId, people } = read();
  return people.find((p) => p.id === activeId) ?? people[0] ?? null;
}

/** Adds or refreshes an identity and makes it the active one. */
export function savePlayer(p: Identity): Store {
  const { people } = read();
  const next = [...people.filter((x) => x.id !== p.id), { ...people.find((x) => x.id === p.id), ...p }];
  const store = { activeId: p.id, people: next };
  write(store);
  return store;
}

export function setActivePlayer(id: string): Store {
  const { people } = read();
  const store = { activeId: people.some((p) => p.id === id) ? id : (people[0]?.id ?? null), people };
  write(store);
  return store;
}

/** Signs this device out of one name, or all of them. */
export function clearPlayer(id?: string): Store {
  const { activeId, people } = read();
  if (!id) {
    write(EMPTY);
    return EMPTY;
  }
  const left = people.filter((p) => p.id !== id);
  const store = { activeId: activeId === id ? (left[0]?.id ?? null) : activeId, people: left };
  write(store);
  return store;
}

File: ios/TallyKit/Sources/TallyKit/Auth/Identity.swift
import Foundation

/// A name this device can pick as, and the token that proves it. Mirrors `src/lib/identity.ts`.
public struct Identity: Codable, Hashable, Sendable, Identifiable {
    public let id: String
    public var name: String
    /// The device token. Absent only for an entry that rides on its account's token.
    public var token: String?
    /// Account whose credential controls this entry (the account's own id for the account itself).
    public var accountId: String?
    /// True when the commissioner put this person on this device.
    public var managed: Bool

    public init(id: String, name: String, token: String? = nil, accountId: String? = nil, managed: Bool = false) {
        self.id = id; self.name = name; self.token = token; self.accountId = accountId; self.managed = managed
    }

    public init(player: Player, token: String? = nil, accountId: String? = nil, managed: Bool = false) {
        self.init(id: player.id, name: player.name, token: token, accountId: accountId, managed: managed)
    }

    public var player: Player { Player(id: id, name: name) }

    /// "A name my account picks for", as opposed to the account itself.
    public var isManagedEntry: Bool {
        if let accountId { return accountId != id }
        return false
    }
}

/**
 Everyone this device can pick as, and who it is picking as right now. Usually one person; more
 when a parent picks for the family. Stored as one Keychain item per host, because an identity
 belongs to Tally on that host, not to a single pool.
 */
public struct SessionStore: Codable, Hashable, Sendable {
    public var activeId: String?
    public var people: [Identity]

    public static let empty = SessionStore(activeId: nil, people: [])

    public init(activeId: String?, people: [Identity]) {
        self.activeId = activeId
        self.people = people
    }

    /// Who this device is picking as right now.
    public var active: Identity? {
        people.first { $0.id == activeId } ?? people.first
    }

    /// The headers every request carries for the active identity.
    public var authHeaders: AuthHeaders {
        guard let p = active else { return AuthHeaders() }
        return AuthHeaders(token: p.token, entryId: p.isManagedEntry ? p.id : nil)
    }

    /// Adds or refreshes an identity and makes it the active one.
    public mutating func save(_ identity: Identity) {
        var merged = identity
        if let existing = people.first(where: { $0.id == identity.id }) {
            if merged.token == nil { merged.token = existing.token }
            if merged.accountId == nil { merged.accountId = existing.accountId }
        }
        people.removeAll { $0.id == identity.id }
        people.append(merged)
        activeId = identity.id
    }

    public mutating func setActive(_ id: String) {
        activeId = people.contains { $0.id == id } ? id : people.first?.id
    }

    /// Signs this device out of one name.
    public mutating func forget(_ id: String) {
        people.removeAll { $0.id == id }
        if activeId == id { activeId = people.first?.id }
    }

    /// Refresh this account's entries without changing which one is selected.
    public mutating func syncAccountEntries(accountId: String, entries: [Player], token: String?) {
        let ids = Set(entries.map(\.id))
        let kept = people.filter { $0.accountId != accountId && !ids.contains($0.id) }
        let fresh = entries.map { Identity(player: $0, token: token, accountId: accountId, managed: $0.id != accountId) }
        people = kept + fresh
        if let activeId, !people.contains(where: { $0.id == activeId }) {
            self.activeId = people.first?.id
        }
    }

    // MARK: Persistence

    private static func key(for host: String) -> String { "session.v1.\(host)" }

    public static func load(host: String, keychain: Keychain = .shared) -> SessionStore {
        guard let data = keychain.data(forKey: key(for: host)),
              let store = try? JSONDecoder().decode(SessionStore.self, from: data) else { return .empty }
        return store
    }

    public func persist(host: String, keychain: Keychain = .shared) {
        if people.isEmpty {
            keychain.remove(forKey: SessionStore.key(for: host))
        } else if let data = try? JSONEncoder().encode(self) {
            keychain.set(data, forKey: SessionStore.key(for: host))
        }
    }
}
